raiden_contracts/contracts/services/UserDeposit.sol

```diff
@@ -0,0 +1,172 @@
+pragma solidity ^0.5.2;
+
+import "raiden/Token.sol";
+import "raiden/Utils.sol";
+
+contract UserDeposit is Utils {
+    string constant public contract_version = "0.5.0";
+    uint constant public withdraw_delay = 100;  // time before withdraw is allowed in blocks
+
+    // Token to be used for the deposit
+    Token public token;
+
+    // Trusted contract (can execute `transfer`)
+    address public msc_address;
+
+    // Keep track of balances
+    mapping(address => uint256) public balances;
+    mapping(address => WithdrawPlan) public withdraw_plans;
+
+    /*
+     *  Structs
+     */
+    struct WithdrawPlan {
+        uint256 amount;
+        uint256 withdraw_block;  // earliest block at which withdraw is allowed
+    }
+
+    /*
+     *  Events
+     */
+
+    event BalanceReduced(address indexed owner, uint newBalance);
+    event WithdrawPlanned(address indexed withdrawer, uint plannedBalance);
+
+    /*
+     *  Modifiers
+     */
+
+    modifier canTransfer() {
+        require(msg.sender == msc_address);
+        _;
+    }
+
+    /*
+     *  Constructor
+     */
+
+    /// @notice Set the default values for the smart contract
+    /// @param _token_address The address of the token to use for rewards
+    constructor(address _token_address)
+        public
+    {
+        // check token contract
+        require(_token_address != address(0x0));
+        require(contractExists(_token_address));
+        token = Token(_token_address);
+        require(token.totalSupply() > 0); // Check if the contract is indeed a token contract
+    }
+
+    /// @notice Specify trusted contracts. This has to be done outside of the
+    /// constructor to avoid cyclic dependencies.
+    /// @param _msc_address Address of the MonitoringService contract
+    function init(address _msc_address)
+        external
+    {
+        // prevent changes of trusted contract after initialization
+        require(msc_address == address(0x0));
+
+        // check monitoring service contract
+        require(_msc_address != address(0x0));
+        require(contractExists(_msc_address));
+        msc_address = _msc_address;
+    }
+
+    /// @notice Deposit tokens. Idempotent function that sets the
+    /// total_deposit of tokens of the beneficiary.
+    /// Can be called by anyone several times and on behalf of other accounts
+    /// @param beneficiary The account benefiting from the deposit
+    /// @param total_deposit The amount tokens in the user's deposit after the call
+    function deposit(address beneficiary, uint256 total_deposit)
```

Comment:
I see an opportunity to explain another kind of common attacks.

1. a user sends a `deposit()` transaction
2. the transaction is seen worldwide in the pending transaction pool
3. MSC (or somebody operating MSC) decides to move tokens out of `balances[beneficiary]`, and sends a `transfer()` call
4. a miner includes MSC's `transfer()` call.  Now `balances[beneficiary]` is smaller than the user thought.
5. a miner includes the user's `deposit()` transaction.  The user pays more than expected.

What is the reason `total_deposit` is used here, instead of just `added_deposit`?

Reply:
The history is related to the reason why went for an idempotent deposit for the core contracts in the first place:  
- https://github.com/raiden-network/raiden-contracts/issues/76
- https://docs.google.com/document/d/1y6jW7yg5Cn6Z5QShQbVDhyVAa4Ogp29k5c45VleZr0I/edit?disco=AAAABp-Q2jo

Reply:
@karlb I reopened https://github.com/raiden-network/raiden-contracts/issues/76 . Will you drive the decision making there?